model/map/src/main/java/org/keycloak/models/map/realm/MapRealmProvider.java

```diff
@@ -0,0 +1,469 @@
+        session.getTransactionManager().enlist(tx);
+    }
+
+    private Function<MapRealmEntity, RealmModel> entityToAdapterFunc() {
+        // Clone entity before returning back, to avoid giving away a reference to the live object to the caller
+
+        return origEntity -> new MapRealmAdapter(session, registerEntityForChanges(origEntity));
+    }
+
+    private MapRealmEntity registerEntityForChanges(MapRealmEntity origEntity) {
+        final MapRealmEntity res = Serialization.from(origEntity);
+        tx.updateIfChanged(origEntity.getId(), res, MapRealmEntity::isUpdated);
+        return res;
+    }
+
+    @Override
+    public RealmModel createRealm(String name) {
+        return createRealm(KeycloakModelUtils.generateId(), name);
+    }
+
+    @Override
+    public RealmModel createRealm(String id, String name) {
+        if (getRealmByName(name) != null) {
+            throw new ModelDuplicateException("Realm with given name exists: " + name);
+        }
+
+        final String entityId = id == null ? KeycloakModelUtils.generateId() : id;
+
+        LOG.tracef("createRealm(%s, %s)%s", id, name, getShortStackTrace());
+
+        MapRealmEntity entity = new MapRealmEntity(entityId);
+        entity.setName(name);
+        if (tx.read(entity.getId()) != null) {
+            throw new ModelDuplicateException("Realm exists: " + entityId);
+        }
+        tx.create(entityId, entity);
+        return entityToAdapterFunc().apply(entity);
+    }
+
+    @Override
+    public RealmModel getRealm(String id) {
+        if (id == null) {
+            return null;
+        }
+
+        LOG.tracef("getRealm(%s)%s", id, getShortStackTrace());
+
+        MapRealmEntity entity = tx.read(id);
+        return entity == null ? null : entityToAdapterFunc().apply(entity);
+    }
+
+    @Override
+    public RealmModel getRealmByName(String name) {
+        if (name == null) return null;
+
+        LOG.tracef("getRealmByName(%s)%s", name, getShortStackTrace());
+
+        ModelCriteriaBuilder<RealmModel> mcb = realmStore.createCriteriaBuilder()
+                .compare(SearchableFields.NAME, Operator.EQ, name);
+
+        String realmId = tx.getUpdatedNotRemoved(mcb)
+                .findFirst()
+                .map(MapRealmEntity::getId)
+                .orElse(null);
+        //we need to go via session.realms() not to bypass cache
+        return realmId == null ? null : session.realms().getRealm(realmId);
+    }
+
+    @Override
+    public Stream<RealmModel> getRealmsStream() {
+        return tx.getUpdatedNotRemoved(realmStore.createCriteriaBuilder())
+                .map(entityToAdapterFunc())
+                .sorted(RealmModel.COMPARE_BY_NAME);
+    }
+
+    @Override
+    public Stream<RealmModel> getRealmsWithProviderTypeStream(Class<?> type) {
+        return StreamUtils.leftInnerJoinStream(tx.getUpdatedNotRemoved(realmStore.createCriteriaBuilder()), MapRealmEntity::getComponents) // create pair of realm and its components
```

Comment:
Can this be done by using ModelCriteriaBuilder?

Reply:
To be honest I haven't tried yet, I'd like to discuss this approach vs MCB approach. I'm not sure what would be advantages  of one over the other. I'd like to hear your opinion, thank you.

Reply:
+1, adding a proper filter would serve the purpose better

Reply:
The advantage is, that currently, you need to download all data from storage and filter it in the stream. However, with MCB, you can filter it on storage level. This means that in case the storage is a database, you will download only entities that fulfill the MCB condition.